kerastuner/engine/oracle.py

```diff
@@ -333,11 +349,37 @@ def _get_oracle_fname(self):
             self._project_dir,
             'oracle.json')
 
-    def _compute_values_hash(self, values):
+    def _compute_values_hash(self, hps):
+        # Don't include conditional params that are not currently active.
+        values = {name: val for name, val in hps.values.items()
+                  if hps.is_active(name)}
         keys = sorted(values.keys())
         s = ''.join(str(k) + '=' + str(values[k]) for k in keys)
         return hashlib.sha256(s.encode('utf-8')).hexdigest()[:32]
 
+    def _random_populate_space(self, allow_collisions=False):
+        for _ in range(self._max_collisions):
+            values = {}
+            for p in self.hyperparameters.space:
+                values[p.name] = p.random_sample(self._seed_state)
+                self._seed_state += 1
+
+            trial_hps = self.hyperparameters.copy()
+            trial_hps.values = values
+            if self._compute_values_hash(trial_hps) in self._tried_so_far:
+                # Attempt to avoid collisions.
+                continue
+            return {'status': trial_lib.TrialStatus.RUNNING,
+                    'values': trial_hps.values}
+
+        # Max collisions occurred.
+        if allow_collisions:
+            return {'status': trial_lib.TrialStatus.RUNNING,
+                    'values': trial_hps.values}
+        # Search space is presumed exhausted.
```

Comment:
The number of collisions is fixed at 20, but 20 is too much if there are 5 possible values and too little if there are 2000.

Perhaps this should be instead a function of the size of the search space (i.e. not an instance attribute but something set at the start of `_random_populate_space`, e.g. 20% of the number of possible values). Exact collisions are only possible for Int, Bool, or Choice, and for each of these the number of values is provided by the constructor params.

Another possibility is to randomly iterate over the values of discrete hps, so that the full space is guaranteed to be covered when it is discrete.

Perhaps this is overly complex, though.